### ISSUE TITLE
Support for cleaning up MapStore/MapLoader instances

### ISSUE THREAD

**Issue by ajermakovics** [hazelcast/hazelcast#4489]
Ported from mailing list https://groups.google.com/forum/#!topic/hazelcast/0_LWfoEMpUE :

> It's currently possible (barely) to write a layer that ensures that MapStore instances (similarly for MapLoader throughout) are cleaned up when their corresponding IMap instances are destroyed. This could be important if the MapStore backing an IMap manages resources that ought to be released when that IMap goes away.

Essentially MapLoaderLifecycleSupport.destroy should be called when IMap is destroyed to close resources

**Comment by Tembrel** [comment]
I'm in favor of this enhancement, but the semantics have to be slightly different, since a MapLoader could be shared by multiple IMap instances. Since `init(hzinst, props, mapName)` is called when a map first starts to use the MapLoader, how about adding `onDestroy(hzinst, mapName)` (or adding a subinterface for this) to be called when an IMap is destroyed, and leave `destroy()` for Hazelcast shutdown?

**Comment by ajermakovics** [comment]
I agree, `onDestroy` or something similar would be useful. Also would make it symmetric with `init` method. Not sure if we need to pass the instance as a parameter since its the same as in `init` but I'll leave the design decisions for later. Subinterface is indeed an option because we need to consider backwards compatibility of the api.
